### PR TITLE
Arrow2D Archetype & Visualizer

### PR DESCRIPTION
**Pull request by Wumpf**
### What

Adds support for 2D arrows.
Lots of copy paste of 3D arrow things, but not so much to be terrible. Was actually pleasantly surprised that I didn't need much codegen-extensions. Deduplication (in particular in the Visualizer) takes more effort than it's worth I figured :)

The full vertical: Archetype, extensions, visualizer roundtrip tests, roundtripable examples!
Only left out dedicated archetype tests since there's nothing new & interesting here really.

That's what the example looks like: (kept it intentionally simple)

image

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/4713/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/4713/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/4713/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4713)
- [Docs preview](https://rerun.io/preview/dc2e334d98665c0f6da6c674b0b0e0cc4e93fbdd/docs) 
- [Examples preview](https://rerun.io/preview/dc2e334d98665c0f6da6c674b0b0e0cc4e93fbdd/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

**Review by jleibs — COMMENTED** [review]
We have an arrows3d_ext.py to enforce positional arguments on the 3d case -- seems like we should have one for the 2d case by symmetry.

File: crates/re_types/definitions/rerun/archetypes/arrows2d.fbs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  // --- Required ---
+
+  /// All the vectors for each arrow in the batch.
+  vectors: [rerun.components.Vector2D] ("attr.rerun.component_required", order: 1000);
+
+  // --- Recommended ---
+
+  /// All the origin (base) positions for each arrow in the batch.
+  ///
+  /// If no origins are set, (0, 0, 0) is used as the origin for each arrow.
```
Comment (jleibs) [inline crates/re_types/definitions/rerun/archetypes/arrows2d.fbs:30]:
```suggestion
  /// If no origins are set, (0, 0) is used as the origin for each arrow.
```

File: rerun_cpp/src/rerun/archetypes/arrows2d_ext.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+// #define EDIT_EXTENSION
+
+namespace rerun {
+    namespace archetypes {
+
+#ifdef EDIT_EXTENSION
+        // <CODEGEN_COPY_TO_HEADER>
+
+        /// Creates new 2D arrows pointing in the given directions, with a base at the origin (0, 0,
+        /// 0).
```
Comment (jleibs) [inline rerun_cpp/src/rerun/archetypes/arrows2d_ext.cpp:12]:
```suggestion
        /// Creates new 2D arrows pointing in the given directions, with a base at the origin (0, 0)
```

File: rerun_cpp/src/rerun/components/vector2d_ext.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            float vector[2];
+#define Vector2D Vector2DExt
+
+            // <CODEGEN_COPY_TO_HEADER>
+
+            /// Construct Vector2D from x/y values.
+            Vector2D(float x, float y) : vector{x, y} {}
+
+            /// Construct Vec2D from x/y/z float pointer.
+            explicit Vector2D(const float* xyz) : vector{xyz[0], xyz[1]} {}
```
Comment (jleibs) [inline rerun_cpp/src/rerun/components/vector2d_ext.cpp:20]:
```suggestion
            /// Construct Vec2D from x/y float pointer.
            explicit Vector2D(const float* xy) : vector{xy[0], xy[1]} {}
```

**Review by jleibs — APPROVED** [review]
Couple of small doc errors but overall looks reasonable. Seems low enough risk I'm fine with slipping it in for 0.12

**Comment by Wumpf** [comment]
good eye on c&p errors. Enforcement of positional arguments is important here as well.
Will address on Monday!

**Comment by nikolausWest** [comment]
So sweet to get in!

**Comment by emilk** [comment]
The example image looks 3D, which is very confusing for something that is explicitly 2D :)
I suggest removing the blue arrow, or pointing it up-right instead

image

**Comment by Wumpf** [comment]
> The example image looks 3D, which is very confusing for something that is explicitly 2D :)

I thought that was a fun twist, but ehrm yeah maybe not helping 😆 
I need something _like_ that arrow otherwise our lack of configurable canvas cuts of the labels of the red & green arrow.
Maybe it's enough to just offset its start a bit? That way the origins aren't always zeroed.

**Comment by Wumpf** [comment]
updated example:

image

**Pull request merged**